### PR TITLE
Create async search index if necessary on updates and deletes

### PR DESCRIPTION
**Pull request by jimczi**
This change ensures that we create the async search index with the right mappings and settings when updating or deleting a document. Users can delete the async search index at any time so we have to re-create it internally if necessary before applying any new operation.

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/user/AsyncSearchUser.java [hunk 1]
```diff
@@ -16,7 +16,8 @@
     public static final AsyncSearchUser INSTANCE = new AsyncSearchUser();
     public static final String ROLE_NAME = UsernamesField.ASYNC_SEARCH_ROLE;
     public static final Role ROLE = Role.builder(new RoleDescriptor(ROLE_NAME,
-            null,
+            // handles template registry for the .async-search index
+            new String[] { "manage_index_templates" },
```
Comment (imotov) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/user/AsyncSearchUser.java:20]: Hmm, does this meant that any async search user can now mess with templates. Does this really necessary?

**Review by imotov — COMMENTED** [review]
LGTM in general, but I wonder if we can somehow avoid extending async search role with index template privilege.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/async/AsyncTaskTemplateRegistry.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    @Override
+    protected List<IndexTemplateConfig> getComposableTemplateConfigs() {
+        return Collections.singletonList(TEMPLATE_ASYNC_SEARCH);
+    }
+
+    @Override
+    protected String getOrigin() {
+        return ClientHelper.INDEX_LIFECYCLE_ORIGIN;
```
Comment (imotov) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/async/AsyncTaskTemplateRegistry.java:56]: It looks like we are reusing this in snapshot as well. So I am ok with this approach. I am just not 100% sure of it needs a separate original and a separate role. Maybe ping somebody from security?

**Review by imotov — APPROVED** [review]

**Comment by jimczi** [comment]
@imotov , we discussed with @albertzaharovits and agreed that an index template is not needed. Instead, we ensure that the index is re-created when necessary on every operation (index, delete and update). We do the same for the security index so that's consistent. I updated the title, description and modified the PR to reflect the last discussions. Can you take another look ?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/async/AsyncTaskIndexService.java [hunk 3]
```diff
@@ -225,7 +227,7 @@ public void deleteResponse(AsyncExecutionId asyncExecutionId,
                                ActionListener<DeleteResponse> listener) {
         try {
             DeleteRequest request = new DeleteRequest(index).id(asyncExecutionId.getDocId());
-            client.delete(request, listener);
+            createIndexIfNecessary(ActionListener.wrap(v -> client.delete(request, listener), listener::onFailure));
```
Comment (imotov) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/async/AsyncTaskIndexService.java:230]: Do we need to create an index in this case or this is for consistency?

**Review by imotov — APPROVED** [review]
I think reasoning for creating an index while deleting response is not very intuitive. Perhaps, a comment explaining a rationale for it could help.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/async/AsyncTaskIndexService.java (hunk 3 above)
Comment (jimczi) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/async/AsyncTaskIndexService.java:236]: I pushed https://github.com/elastic/elasticsearch/pull/64606/commits/a273d920cd9f91e67507abe2c1c321c3e9eb62ad to explain the reasoning.

Comment (imotov) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/async/AsyncTaskIndexService.java:236]: :+1:

**Pull request merged**